### ISSUE TITLE
Multithreaded Benchmarks

### ISSUE THREAD

**Issue by mrks** [hyrise/hyrise#1005]
The benchmark runner is a good tool to evaluate our performance. Currently, it only executes one query at a time (which might be executed with multiple threads). @lanice built some [cool graphs](https://github.com/hyrise/hyrise/pull/994#issuecomment-408680431) for multi-threaded multi-client benchmarks.

It should be possible to run this as easily as the current benchmarks are and get a similar graph. Because this only makes sense with a plot, I think we should not embed this in the runner, but have a wrapper script for the runner.

Interesting command line options:
- number of clients (i.e., queries running in parallel)
- number of queries / max execution time
- thread stepping (for a quick evaluation, it should be enough to evaluate 1, 2, 3, 4, 5, 10, 15, ... threads)

Before we start to work on this, let's discuss it.

**Comment by Bouncner** [comment]
To get results without any NUMA effects (I am obviously sure that NUMA is in excellent hands though), it would be nice to have means to restrict Hyrise to n CPU sockets.

Is that alreadt implicitly handles via the cores? Or would this be something new? (I guess it’s something new?)

**Comment by lanice** [comment]
Yes, a few thoughts on that:
For my benchmarks, I did embed it into the benchmark runner code (read: hijacked the runner for my purposes), because of what you said, the runner runs each query individually, and I wanted to start multiple runs per query simultaneously.
Starting multiple clients through a script could be possible as well, although I haven't run any tests like that, and also I'd imagine the results get pretty messy. Running 500 clients at once, each one spitting out a result file for each core tested...
Wouldn't it be more practical to alter the benchmark runner to support running multiple queries in parallel?
Besides that, I think having a wrapper script in general is a good idea, it's what I am doing right now anyway.

**Comment by lanice** [comment]
@Bouncner Yes, it is handled implicitly. `Topology::use_numa_topology(N)` creates a NUMA topology with N threads, filling up one socket after another.  `Topology::use_numa_topology(20)` on a 4-socket (20 threads per socket) machine uses exactly each thread on the first socket, and so on.

But I could imagine being able to specify the number of sockets instead of threads as well, to tell the Topology "use all available threads on X sockets".

**Comment by mrks** [comment]
>  Wouldn't it be more practical to alter the benchmark runner to support running multiple queries in parallel?

Might be easier, yes. You have more experience with that benchmark.

I think @Bouncner was asking if `20` would mean that only the memory from the first node would be used (e.g., via numactl)

**Comment by mrks** [comment]
From our meeting:
* Offer parameter `--clients=x` (or `parallel_queries`, ...)
* Used both in `_benchmark_permuted_query_sets` and `_benchmark_individual_queries`
* SQL Parsing/Optimization should be scheduled as well
* How to we measure the elapsed time / time per iteration if one client returns significantly later than the others?
